refactor(docs-infra): fix template eslint issues in aio examples

dario-piotrowicz opened the pull request:
the aio examples have various eslint issues regarding template rules, those
are currently turned off and TODO comments have been added to them in the
examples eslintrc, fix such issues and remove the respective TODO comments

this also includes examples refactoring to use buttons for better accessibility,
this change tries to make the smallest amound of changes to the examples' behaviors
and designs/UI

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [x] angular.io application / infrastructure changes
- [ ] Other... Please describe:


## Issue
Issue Number: N/A

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No


## Other information
@gkalpak 😃 👍

dario-piotrowicz committed:
refactor(docs-infra): fix template eslint issues in aio examples

the aio examples have various eslint issues regarding template rules, those
are currently turned off and TODO comments have been added to them in the
examples eslintrc, fix such issues and remove the respective TODO comments

this also includes examples refactoring to use buttons for better accessibility,
this change tries to make the smallest amound of changes to the examples' behaviors
and designs/UI

dario-piotrowicz commented on aio/content/examples/animations/src/app/hero-list-page.component.css:
```diff
@@ -1,43 +1,74 @@
   font-size: small;
   color: white;
   padding: 0.8em 0.7em 0 0.7em;
-  background-color: #3d5157;
-  position: relative;
-  left: -1px;
-  top: -4px;
-  height: 1.8em;
-  min-width: 16px;
-  text-align: right;
```
I made this change here, but in other instances I've kept the relative positioning 🤔 

probably I should update them all in the same way for consistency sake

(unfortunately they are not all like 100% the same... it's such a shame we have such code repetition around the heroes list 😓 )

dario-piotrowicz replied:
I've updated most of the heroes lists to follow this pattern (instead of a relative positioning and negative left and top, I've defined a flex container and used that to position the badge)

not all use heroes list use this pattern and not all are identical, but most of them are very similar, so hopefully this should be ok consistency wise 🙂

dario-piotrowicz commented on aio/content/examples/animations/src/app/hero-list-page.component.css:
```diff
@@ -1,43 +1,74 @@
-  height: 1.8em;
-  min-width: 16px;
-  text-align: right;
+  background-color:#405061;
+  line-height: 1em;
```
I copy pasted this block of code from somewhere else, I kept the backround-color and the line-height from there to have a more consisteny styling 🙂 

(ps: missing space after the background-color colon, need to fix it)

dario-piotrowicz commented on aio/content/examples/docs-style-guide/src/app/app.component.css:
```diff
@@ -6,8 +6,19 @@
 }
 /* #enddocregion heroes */
-
+/* HeroesComponent's private CSS styles */
```
to remove

dario-piotrowicz commented on aio/content/examples/router/src/app/admin/admin-dashboard/admin-dashboard.component.2.html:
```diff
@@ -1,5 +1,4 @@
 <h3>Dashboard</h3>
 
 <p>Session ID: {{ sessionId | async }}</p>
-<a id="anchor"></a>
```
I could not figure out why this was here and what for 😓 , does someone know? 🙂

dario-piotrowicz commented on aio/content/examples/router/src/app/admin/admin-dashboard/admin-dashboard.component.html:
```diff
@@ -1,7 +1,6 @@
 <h3>Dashboard</h3>
 
 <p>Session ID: {{ sessionId | async }}</p>
-<a id="anchor"></a>
```
I could not figure out why this was here and what for 😓 , does someone know? 🙂

dario-piotrowicz replied:
ah ok sorry I missed that :sweat: , I'll revert this change then :slightly_smiling_face:

dario-piotrowicz replied:
I've putted it back but as a `div` :slightly_smiling_face:

dario-piotrowicz commented on aio/content/examples/toh-pt2/src/app/heroes/heroes.component.html:
```diff
@@ -2,10 +2,11 @@
+  <li *ngFor="let hero of heroes">
+    <button [class.selected]="hero === selectedHero"
+      (click)="onSelect(hero)">
+      <span class="badge">{{hero.id}}</span> {{hero.name}}
```
missing indentation 😓

dario-piotrowicz commented on aio/content/examples/toh-pt3/src/app/heroes/heroes.component.html:
```diff
@@ -1,10 +1,11 @@
+  <li *ngFor="let hero of heroes">
+    <button [class.selected]="hero === selectedHero"
+      (click)="onSelect(hero)">
+      <span class="badge">{{hero.id}}</span> {{hero.name}}
```
missing indentation 😓

dario-piotrowicz commented:
Note: for `img` with missing `alt`, I’ve added those most of the time, hopefully it doesn’t make the examples less clear (if you think it does please let me know and I will just add eslint disable comments instead)

I’ve just disable those eslint errors in a couple of places where I felt that adding an `alt` would make the example less clear

dario-piotrowicz commented on aio/content/examples/architecture/src/app/hero-list.component.1.html:
```diff
@@ -1,6 +1,7 @@
 <li>{{hero.name}}</li>
 <app-hero-detail [hero]="selectedHero"></app-hero-detail>
+<!-- eslint-disable-next-line @angular-eslint/template/click-events-have-key-events -->
 <li (click)="selectHero(hero)"></li>
```
I'm not completely sure what this is for.... so I figured I could just ignore the line to be on the safe side 😓

dario-piotrowicz replied:
yeah I know understand the numbering in the filenames :sweat_smile: 

this doesn't align with the full file so I need to update it :+1:

dario-piotrowicz commented on aio/content/examples/toh-pt2/src/app/heroes/heroes.component.1.html:
```diff
@@ -12,6 +12,7 @@ <h2>My Heroes</h2>
 
 <!-- #docregion selectedHero-click -->
+<!-- eslint-disable-next-line @angular-eslint/template/click-events-have-key-events -->
 <li *ngFor="let hero of heroes" (click)="onSelect(hero)">
```
I'm not completely sure what this is for.... so I figured I could just ignore the line to be on the safe side 😓

dario-piotrowicz commented:
@gkalpak just wanted to ask you, I had to change some protractor tests in this PR, this made me wonder, since protractor is on its way out, should be maybe look into re-implementing the e2es with some other tool? 🙂 

> I would love to help in that if I can 🙂  (btw I've got Cypress experience and I do think it is a great tool nice to work with 😄 )

mary-poppins commented:
You can preview 695fd55 at https://pr44557-695fd55.ngbuilds.io/.
You can preview 6ba979d at https://pr44557-6ba979d.ngbuilds.io/.
You can preview 9668614 at https://pr44557-9668614.ngbuilds.io/.
You can preview 770211b at https://pr44557-770211b.ngbuilds.io/.
You can preview ae5eacb at https://pr44557-ae5eacb.ngbuilds.io/.
You can preview 31bd056 at https://pr44557-31bd056.ngbuilds.io/.
You can preview d1f792b at https://pr44557-d1f792b.ngbuilds.io/.
You can preview a6cdc2b at https://pr44557-a6cdc2b.ngbuilds.io/.
You can preview 441f4b8 at https://pr44557-441f4b8.ngbuilds.io/.
You can preview 5201481 at https://pr44557-5201481.ngbuilds.io/.
You can preview 0f2126d at https://pr44557-0f2126d.ngbuilds.io/.
You can preview cc64142 at https://pr44557-cc64142.ngbuilds.io/.
You can preview 2bac073 at https://pr44557-2bac073.ngbuilds.io/.
You can preview c4195a6 at https://pr44557-c4195a6.ngbuilds.io/.
You can preview ba6fcc1 at https://pr44557-ba6fcc1.ngbuilds.io/.
You can preview 7c0971b at https://pr44557-7c0971b.ngbuilds.io/.

dylhunn reviewed (approved):
This is too large for me to review all changes line-by-line, but the general impact looks really good. Let me know if you'd like eyes on anything specific in detail. You'll still need other approvals for other areas.

reviewed-for: fw-core, fw-forms, fw-http, fw-platform-server, fw-testing, docs-observables

dario-piotrowicz commented:
> This is too large for me to review all changes line-by-line, but the general impact looks really good. Let me know if you'd like eyes on anything specific in detail. You'll still need other approvals for other areas.
> 
> reviewed-for: fw-core, fw-forms, fw-http, fw-platform-server, fw-testing, docs-observables

Thank you so very much! 😄 

I think it should be all ok, but since you're asking 😜 , if possible I would love it if you could pick randomly a couple of examples in which I switched `label` to `span` and double check that the example (in stackblitz should be enough I think) does still look good (hopefully exactly the same), I did so for a couple of them, if anyone could also do the same for another random couple that'd be great 😄  (but I did check the code and it should indeed be fine, unless there's some quite hidden(/up the chain css) that targets specifically labels that I've missed)

jessicajaniuk commented on aio/content/examples/template-syntax/src/app/app.component.html:
```diff
@@ -221,8 +221,10 @@ <h3>
 
 <app-hero-detail prefix="You are my" [hero]="currentHero"></app-hero-detail>
 
+<!-- eslint-disable @angular-eslint/template/accessibility-alt-text -->
```
Rather than disabling alt text here, can we just create some that's relevant? Like `interpolated image of {{currentHero.name}}` would work.

jessicajaniuk commented on aio/content/examples/template-syntax/src/app/app.component.html:
```diff
@@ -181,6 +180,7 @@ <h3>
 
 <a class="to-toc" href="#toc">top</a>
 
+<!-- eslint-disable @angular-eslint/template/accessibility-alt-text -->
```
Can we just add some alt text here?

dario-piotrowicz replied:
I am afraid that would just make the examples more confusing

I am not totally sure what the first example is showing, but it does seem to be specific to only `src` so adding an `alt` would just risk to make things less clear

the second example shows the difference between the property binding `[src]` and the attribute binding `[attr.src]` here even more I believe that throwing alts in the mix would make things more confusing (we could add `[alt]` and `[attr.alt]` but it doesn't really seem very clear to me what's going on)

so I am not really sure, I would prefer not put `alt`s here, what do you think? :sweat: 

Also, alternatively if we want to bind a single property/attribute I think that the element doesn't actually need to necessarily be the an `img`, for example instead of having images we could have an `input` and compare `[type]` to `[attr.type]`, couldn't we? :thinking:

jessicajaniuk replied:
It depends. If we're actually displaying the image, and not just showing demo code, it may be fine. However skipping alt text violates WCAG accessibility rules. So I'd prefer we have them. We should be demonstrating a11y best practices. So we should have alt text for real images and those alt texts should be clear as to what's in the image disambiguating it from other images and providing context.

dario-piotrowicz replied:
This is just demo code, there isn't actually any valid image assigned to those `img`s as you can see from the example, which you can find here: [template syntax example](https://angular.io/guide/example-apps-list#template-syntax)

this anyways is that it looks like:
![Screenshot at 2022-01-19 20-11-55](https://user-images.githubusercontent.com/61631103/150206637-2f225924-9d37-4202-8c66-ef9efa06c081.png)

So I would actually use different elements with different properties/attributes to try to keep this as similar as possible to the existing example but without violating accessibility rules, what do you think? :slightly_smiling_face:

jessicajaniuk replied:
I think as long as we are doing the alt texts in a way where inside these examples it's distinguishable between images, then I'm ok with it.

jessicajaniuk replied:
Sorry, I missed the last bit there and was focused on the img code. Yes, we can definitely use different elements.

jessicajaniuk replied:
I think I'm just going to say let's move forward as is. I think given where we were before and where we are now is markedly better than we were. We can increment and add effective alt text later.

dario-piotrowicz replied:
sure :slightly_smiling_face: 

I wanted to check with @petebacondarwin (which I think created the example there) what he thinks about switching to a different HTML element (as in, there is a specific reason/benefit as to why `img`s were used?)

Anyways @jessicajaniuk as you said that can also come as a follow up PR in case :slightly_smiling_face:

jessicajaniuk commented on aio/content/examples/property-binding/src/app/app.component.html:
```diff
@@ -47,8 +47,8 @@ <h3>Pass objects:</h3>
   <!-- #docregion property-binding-interpolation -->
-  <p><img src="{{itemImageUrl}}"> is the <i>interpolated</i> image.</p>
-  <p><img [src]="itemImageUrl"> is the <i>property bound</i> image.</p>
+  <p><img alt="item" src="{{itemImageUrl}}"> is the <i>interpolated</i> image.</p>
```
If we know the item, can we make sure these two alt tags are unique? We could put something like `item {{item.name}}` interpolated image`. I don't know if that property is there, but that ensure we don't have 3 images with `item` as the alt text.

dario-piotrowicz replied:
the `itemImageUrl` is always the same, so it doesn't make sense for the `alt` to be the same as well? (since the image is always supposed to be the same) :thinking:

jessicajaniuk replied:
What it's rendering is implying different things here. So I think you can distinguish alt text with "Interpolated item image" and "Property Bound item image" to distinguish between the two in the example. It may be the same URL but the purpose of them is different, and the alt text should convey that.

dario-piotrowicz replied:
ah ok sure I can go with that :slightly_smiling_face: (without the `Image` part I think because that gets already dictated :thinking:)

dario-piotrowicz replied:
I made the change I hope I got it right :smiley:

jessicajaniuk commented on aio/content/examples/interpolation/src/app/app.component.html:
```diff
@@ -35,7 +35,7 @@ <h2>Expression Context</h2>
   <!-- #docregion component-context -->
   <h4>{{recommended}}</h4>
-  <img [src]="itemImageUrl2">
+  <img alt="item" [src]="itemImageUrl2">
```
We should ensure that two images on the same page don't have the same alt text.

dario-piotrowicz replied:
I just put `"item 2"` as the alt, I hope it makes enough sense given that the image url ends with 2 :slightly_smiling_face:

jessicajaniuk reviewed (changes requested):
Overall this looks great, @dario-piotrowicz. My only concern is the alt tags. Can you address those?

dario-piotrowicz committed:
fixup! refactor(docs-infra): fix template eslint issues in aio examples

mary-poppins commented:
You can preview fad8f03 at https://pr44557-fad8f03.ngbuilds.io/.

dario-piotrowicz committed:
fixup! refactor(docs-infra): fix template eslint issues in aio examples

mary-poppins commented:
You can preview 358151a at https://pr44557-358151a.ngbuilds.io/.

gkalpak commented on aio/content/guide/binding-syntax.md:
```diff
@@ -237,7 +237,7 @@ The following table summarizes the targets for the different binding types.
     </td>
     <td>
-      <code>src</code>, <code>hero</code>, and <code>ngClass</code> in the following:
+      <code>alt</code>, <code>src</code>, <code>hero</code>, and <code>ngClass</code> in the following:
```
:100:

gkalpak commented on aio/content/examples/interpolation/src/app/app.component.html:
```diff
@@ -11,7 +11,7 @@ <h3>Current customer: {{ currentCustomer }}</h3>
     <!-- #docregion component-property -->
     <p>{{title}}</p>
-    <div><img src="{{itemImageUrl}}"></div>
+    <div><img alt="item" src="{{itemImageUrl}}"></div>
```
Nit: Maybe use "The first item."/"The second item." as alt text?

dario-piotrowicz replied:
mh... I am not completely sure, the fact is that they are a bit far apart ([aio link](https://angular.io/guide/interpolation#displaying-values-with-interpolation))

so when someone first reads the `Displaying values with interpolation` they do not see the two items there and could look a bit strange to just see a "The first item" without a second one

But that may just be my opinion, I can use your proposed alt text if you do prefer that version :slightly_smiling_face:

gkalpak replied:
I don't feel strongly about it, so up to you.

dario-piotrowicz replied:
I don't feel strongly about it either :stuck_out_tongue_winking_eye: , but I do slightly prefer it with the number, so I guess if you're ok with it I'll keep it as it is :slightly_smiling_face:

(unless @jessicajaniuk thinks otherwise :slightly_smiling_face:)

gkalpak reviewed (changes requested):
Amazing work getting all these errors fixed, @dario-piotrowicz 🥇
I just left a few (mostly minor) comments.

The CSS duplication is indeed not great, but I don't think we can do much about it (and definitely not as part of this PR - the duplication was there before). FWIW, we have a [common stylesheet][1] that is included in all examples, but I am not sure the style changes in this PR are generic enough to be included there.

[1]: https://github.com/angular/angular/blob/af2a1317cbfba985233f510bbe9e49016968307c/aio/tools/examples/shared/boilerplate/common/src/styles.css

dario-piotrowicz commented on aio/content/examples/animations/src/app/hero-list-groups.component.ts:
```diff
@@ -33,15 +32,15 @@ import { Hero } from './hero';
     trigger('flyInOut', [
       state('in', style({
-        width: 120,
+        width: '*',
```
it was to fix some small visual issue if I remember correctly :thinking:

I'll look into that and give you more details :slightly_smiling_face:

dario-piotrowicz replied:
Ok, `120` is an hardcoded value and doesn't work very well, this is how the thing actually currently looks in the aio live example:
![currently](https://user-images.githubusercontent.com/61631103/149019114-321f10ec-cab3-4842-a19f-abee14ef607e.gif)
as you can see using `*` (which relies on the current width of the element instead of providing a hardcoded value) works much better

Furthermore I am actually updating the styles in order to make the various heroes lists consistent looking so with my changes this is what the thing looks like:
![withmychange](https://user-images.githubusercontent.com/61631103/149019410-c9418d7d-c19c-4c88-a029-1e55393cfde3.gif)

:slightly_smiling_face:

gkalpak replied:
Thx for the detailed comment :100:

dario-piotrowicz commented on aio/content/examples/docs-style-guide/src/app/app.component.html:
```diff
@@ -1,17 +1,18 @@
 <div *ngIf="selectedHero">
   <h2>{{selectedHero.name}} details!</h2>
-  <div><label>id: </label>{{selectedHero.id}}</div>
+  <div><span>id: </span>{{selectedHero.id}}</div>
```
yeah good point! :stuck_out_tongue_winking_eye:

dario-piotrowicz replied:
PS: below it should actually be a `label` associated to that input :slightly_smiling_face:

dario-piotrowicz commented on aio/content/examples/hierarchical-dependency-injection/src/app/hero-tax-return.component.html:
```diff
@@ -2,15 +2,15 @@
   </fieldset>
   <fieldset>
-    <label>
+    <span>
```
yeah.... doesn't seem right, I may have just changed it by mistake with the rest present in the file, sorry! :sweat: 

I'll double check it and revert it back :slightly_smiling_face::+1:

dario-piotrowicz commented on aio/content/examples/toh-pt2/src/app/heroes/heroes.component.1.html:
```diff
@@ -12,6 +12,7 @@ <h2>My Heroes</h2>
 <!-- #enddocregion li -->
 
 <!-- #docregion selectedHero-click -->
+<!-- eslint-disable-next-line @angular-eslint/template/click-events-have-key-events -->
```
ahhh, I didn't know! :sweat: 

thanks so much, I'll double check and make sure they do match :slightly_smiling_face::+1:

dario-piotrowicz replied:
yeah I had it quite differently, I've fixed this so that it looks like the final version of the file and updated the aio toh guide accordingly :slightly_smiling_face: 

(PS: in the toh I am saying to add a `button` with the click event binding instead of saying to add the event binding to the `li` element, I would like to also add an alert or callout saying that for accessibility it is generally better to use `button` elements instead of having `click` event bindings assigned to other elements, but that may be a bit out of place in the toh, what do you think?)

gkalpak replied:
I'm on the fence on that too. Whatever you think is best 😅

dario-piotrowicz replied:
I've tried adding an alert ([link](https://pr44557-633bb35.ngbuilds.io/tutorial/toh-pt2#add-a-click-event-binding)) I am not totally convinced.... what do you think? :confused:

dario-piotrowicz commented on aio/content/examples/testing/src/app/hero/hero-list.component.spec.ts:
```diff
@@ -68,7 +68,7 @@ describe('HeroListComponent', () => {
 
        // In older browsers, such as IE, you might need a CustomEvent instead. See
        // https://developer.mozilla.org/en-US/docs/Web/API/CustomEvent/CustomEvent#Polyfill
```
@gkalpak I think this comment can be removed ([caniuse Event](https://caniuse.com/mdn-api_event_event)) what do you think? :slightly_smiling_face:

dario-piotrowicz replied:
(same for the same comment above)

gkalpak replied:
Agreed :+1:

dario-piotrowicz committed:
fixup! refactor(docs-infra): fix template eslint issues in aio examples

mary-poppins commented:
You can preview 1ad40aa at https://pr44557-1ad40aa.ngbuilds.io/.

dario-piotrowicz committed:
fixup! refactor(docs-infra): fix template eslint issues in aio examples

dario-piotrowicz commented on aio/content/examples/upgrade-phonecat-1-typescript/app/phone-detail/phone-detail.template.html:
```diff
@@ -1,5 +1,5 @@
 <div class="phone-images">
-  <img ng-src="{{img}}" class="phone"
+  <img ng-src="{{img}}" class="phone" alt="phone {{$index}}"
```
@gkalpak I am not too sure what to use here as `alt`s :confused: 

this first list and the one below seem to be of the same images (for the same phone), I think the first just shows all the images and styles the selected one differently

below instead the same list is used to actually select the image.... `phone {{$index}}` and `phone thumb {{$index}}` aren't really appropriate.... but do you have any suggestion?

PS: since the images are the same I guess also the alts should be the same, maybe something like: `phone view ${{index}}`? how does that sound? :slightly_smiling_face:

dario-piotrowicz replied:
( Here and in the files below too of course :slightly_smiling_face: )

gkalpak replied:
Couldn't we use the phone's name here?

dario-piotrowicz replied:
I don't think so :sweat: 

this is the application on a random phone I just picked: [phonecat motorola](http://angular.github.io/angular-phonecat/step-14/app/#!/phones/motorola-xoom-with-wi-fi)

As you can see:
![Screenshot at 2022-01-12 19-53-19](https://user-images.githubusercontent.com/61631103/149211914-231d26ee-fbe7-4680-b260-bb726f292b5e.png)

The images are all of the same phone so it wouldn't make much sense to just use the same exact `alt` (the phone's name) for all of them, right? :thinking:

gkalpak replied:
Yeah, sorry, I meant use the phone's name in the alt text. Something like "Phone X - thumbnail Y".

dario-piotrowicz replied:
sounds good, thanks :slightly_smiling_face::+1:

mary-poppins commented:
You can preview 64fc9b0 at https://pr44557-64fc9b0.ngbuilds.io/.

dario-piotrowicz committed:
fixup! refactor(docs-infra): fix template eslint issues in aio examples

mary-poppins commented:
You can preview 791e02c at https://pr44557-791e02c.ngbuilds.io/.

mary-poppins commented:
You can preview 132c4a3 at https://pr44557-132c4a3.ngbuilds.io/.

gkalpak commented on aio/content/examples/router/src/app/admin/admin-dashboard/admin-dashboard.component.2.html:
```diff
@@ -1,5 +1,5 @@
 
 <p>Session ID: {{ sessionId | async }}</p>
-<a id="anchor"></a>
+<div id="anchor"></div>
```
This whole file doesn't seem to be used, so I think it can be removed altogether.

dario-piotrowicz replied:
ah wow, you're right! :open_mouth::+1:

dario-piotrowicz committed:
refactor(docs-infra): remove outdated comments regarding CustomEvent

remove the comments present in aio examples saying that for older
browsers such as IE CustomEvent could be necessary (as the the Event
constructor could be not available)

such comments are no longer necessary/relevant since the Event constructor
is available for all the supported browsers

dario-piotrowicz committed:
fixup! refactor(docs-infra): fix template eslint issues in aio examples

mary-poppins commented:
You can preview 72173ae at https://pr44557-72173ae.ngbuilds.io/.

dario-piotrowicz committed:
fixup! refactor(docs-infra): fix template eslint issues in aio examples

mary-poppins commented:
You can preview 633bb35 at https://pr44557-633bb35.ngbuilds.io/.

dario-piotrowicz committed:
fixup! refactor(docs-infra): fix template eslint issues in aio examples

dario-piotrowicz committed:
fixup! refactor(docs-infra): fix template eslint issues in aio examples

dario-piotrowicz commented on aio/content/guide/architecture-components.md:
```diff
@@ -89,14 +89,14 @@ This example from the `HeroListComponent` template uses three of these forms.
 
 * The `(click)` [*event binding*](guide/user-input#binding-to-user-input-events) calls the component's `selectHero` method when the user clicks a hero's name.
 
+* The `{{hero.name}}` [*interpolation*](guide/interpolation)
+displays the component's `hero.name` property value within the `<button>` element.
+
```
I changed this to keep the order or the binding in the code and in the list

before:
[![Screenshot at 2022-01-12 21-54-24](https://user-images.githubusercontent.com/61631103/149229056-212f9c5a-d231-47fa-914e-37ee2f0220f4.png)](https://angular.io/guide/architecture-components#data-binding)

after:
[![Screenshot at 2022-01-12 21-53-43](https://user-images.githubusercontent.com/61631103/149229068-6d07e7a3-4383-4e46-aff4-9de80325f1a5.png)](https://pr44557-fc99d3f.ngbuilds.io/guide/architecture-components#data-binding)

I think it is a bit uglier now, but this was needed in order to keep the button :sweat: 

(alternatively I can just put the eslint disable comment for the li click binding)

@gkalpak what do you think? :slightly_smiling_face: :sweat:

gkalpak replied:
I think this seems reasonable. The two `<li>` didn't make much sense (esp. since the second is empty).

dario-piotrowicz replied:
yeah :slightly_smiling_face: 

it was nicer to have 3 lines and 3 bullet points, but hopefully this is just as clear :smile:

gkalpak commented on aio/content/tutorial/toh-pt2.md:
```diff
@@ -117,15 +117,31 @@ and update the hero detail.
 
+<div class="callout is-helpful">
+
+<header>Clickable Elements</header>
```
```suggestion
  <header>Clickable elements</header>
```

gkalpak commented on aio/content/tutorial/toh-pt2.md:
```diff
@@ -117,15 +117,31 @@ and update the hero detail.
+
+<header>Clickable Elements</header>
+
+  In order to add a click event binding we've added a `<button>` element to our `<li>`, note that this
+  is not the only option, and we could have just as easily added the click event binding directly to the
+  `<li>` element.
+
+  We can add the click event binding to any possible element, but this is generally not advised as doing
+  so hinders assistive technologies and prevents automatic keyboard behaviors (e.g. the click being emitted
+  also when the user presses the enter key).
+
+  These problems can still be overcome by adding additional event bindings and ARIA attributes, but by using
+  `<button>` elements we get all the desired behaviors for free.
```
I suggest keeping this shorter and focused to clicks (since it is a side-note and targets beginners). How about:

```suggestion
  Note that we added the click event binding on a new `<button>` element. While we could have added
  the event binding on the `<li>` element directly, it is better for accessibility purposes to use
  the native `<button>` element to handle clicks.

  For more details on accessibility, see [Accessibility in Angular](guide/accessibility).
```

gkalpak reviewed (changes requested):
One last comment from me 🚀

dario-piotrowicz committed:
fixup! refactor(docs-infra): fix template eslint issues in aio examples

mary-poppins commented:
You can preview 38d6476 at https://pr44557-38d6476.ngbuilds.io/.

dario-piotrowicz committed:
fixup! refactor(docs-infra): fix template eslint issues in aio examples

dario-piotrowicz commented:
@gkalpak should be all good now :smile: 

as far as I can tell there are only a couple of comments left to address with @jessicajaniuk :slightly_smiling_face:

mary-poppins commented:
You can preview 826440d at https://pr44557-826440d.ngbuilds.io/.

gkalpak commented on aio/content/examples/upgrade-phonecat-2-hybrid/app/phone-detail/phone-detail.template.html:
```diff
@@ -1,7 +1,7 @@
 <div *ngIf="phone">
   <div class="phone-images">
-    <img [src]="img" class="phone" [alt]="'phone ' + index"
+    <img [src]="img" class="phone" [alt]="'Phone ' + phone.name + ' - thumbnail ' + index"
```
Nit: I find that the attribute binding would be a little more readable here (and below):

```suggestion
    <img [src]="img" class="phone" alt="Phone {{ phone.name }} - thumbnail {{ index }}"
```

gkalpak commented on aio/content/tutorial/toh-pt2.md:
```diff
@@ -130,16 +130,10 @@ When the user clicks in the `<button>`, Angular executes the `onSelect(hero)` ex
+  Note that we added the click event binding on a new `<button>` element. While we could have added
+  the event binding on the `<li>` element directly, it is better for accessibility purposes to use
+  the native `<button>` element to handle clicks.
+  For more details on accessibility, see [Accessibility in Angular](guide/accessibility).
```
Nit: I'd leave an empty line above this sentence.

dario-piotrowicz replied:
sorry I must removed it by mistake as I copy-pasted your suggestion :sweat:

gkalpak reviewed (approved):
:tada:

Thx for making all the changes, @dario-piotrowicz!
Amazing work 🚀 :shipit:

gkalpak commented on aio/content/tutorial/toh-pt2.md:
```diff
@@ -117,15 +117,25 @@ and update the hero detail.
 
+<div class="callout is-helpful">
+
+<header>Clickable Elements</header>
```
```suggestion
  <header>Clickable elements</header>
```

dario-piotrowicz replied:
sorry I was pretty sure I already made this change! :bow: :cry:

dario-piotrowicz committed:
fixup! refactor(docs-infra): fix template eslint issues in aio examples

gkalpak reviewed (approved):
:shipit:

mary-poppins commented:
You can preview 5f9125f at https://pr44557-5f9125f.ngbuilds.io/.

mary-poppins commented:
You can preview 5436947 at https://pr44557-5436947.ngbuilds.io/.

jessicajaniuk commented on aio/content/examples/http/src/app/heroes/heroes.component.html:
```diff
@@ -14,15 +14,15 @@ <h2>Heroes</h2>
 <ul class="heroes">
   <li *ngFor="let hero of heroes">
-    <a (click)="edit(hero)">
+    <button class="edit" (click)="edit(hero)">
```
It's invalid HTML to put a text input inside a button. This will need to be re-structured to make it valid.

dario-piotrowicz replied:
Yes and it is slightly awkward from the UI as well

You can check the example here: [server communication example](https://angular.io/guide/example-apps-list#server-communication)

I'd imagine that as it is (an `input` inside an `a`) is just as invalid, so at the very least I am not making things any worse.

I am wondering, would it be ok to keep this as is for this PR and do a more proper rework/cleanup of the example in a separate one? (as I am seeing that this is not the only thing that could be improved there)

jessicajaniuk replied:
You are correct in that an input inside an anchor tag is just as invalid. So a follow up is fine given the effort. Thanks!

jessicajaniuk commented on aio/content/examples/animations/src/app/hero-list-auto.component.html:
```diff
@@ -1,9 +1,9 @@
-      [@shrinkOut]="'in'" (click)="removeHero(hero.id)">
-      <div class="inner">
+      [@shrinkOut]="'in'">
+      <button class="inner" (click)="removeHero(hero.id)">
```
Any time we have a button, we should be defining a type attribute. Can we just mass add `type="button"` to all of these?

dario-piotrowicz replied:
I've added the `type` to all the buttons in the aio examples :smile: 

I had to add them one by one as I needed to check that they didn't already have one and also make sure that the type made sense (there were a couple of times where I was tempted to use the `"reset"` type but at the end of the day those were normal buttons too, just with a sort of reset-like logic)

you can check out the commit :slightly_smiling_face: 
(I made it as a separate commit and not a fixup one as it feels a bit out of the scope of the original one, I hope that's ok :slightly_smiling_face:)

jessicajaniuk reviewed (changes requested):
Almost there! Just a couple of things.

dario-piotrowicz committed:
refactor(docs-infra): add types to all the aio examples buttons

add types to all the buttons in the angular.io examples to encourage
the best practice of always including a type per button (regardless
to whether it is in a form or now)

mary-poppins commented:
You can preview 899120c at https://pr44557-899120c.ngbuilds.io/.

mary-poppins commented:
You can preview 5f50abb at https://pr44557-5f50abb.ngbuilds.io/.

dario-piotrowicz commented:
sorry @dylhunn I re-requested your review by mistake :sweat: :bow:

dario-piotrowicz commented:
@jessicajaniuk please have another look, most of your comments should be addressed :slightly_smiling_face: 

I think we just need to:
 - sort this conversation out: https://github.com/angular/angular/pull/44557#discussion_r779236631
 - decide if a follow up PR is ok for this: https://github.com/angular/angular/pull/44557#discussion_r787239658

mary-poppins commented:
You can preview 64680e5 at https://pr44557-64680e5.ngbuilds.io/.

dario-piotrowicz committed:
fixup! refactor(docs-infra): fix template eslint issues in aio examples

jessicajaniuk reviewed (approved):
LGTM :cookie:

Thanks for your work, Dario!

jessicajaniuk reviewed (approved):
reviewed-for: public-api

AndrewKushnir reviewed (approved):
Reviewed-for: public-api

jessicajaniuk commented:
Caretaker Note: We can ignore the public-api reviews. This PR affects files that are in pull approve for public-api, but just the examples change. Nothing actually changes in the public api.

dylhunn commented:
This PR was merged into the repository by commit 0d19be199617b0bf2f65b19bf95a41247978fb8d.
